File: 0155-min-stack/solution.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.min = []
        
    def push(self, x: int) -> None:
        if len(self.stack) == 0:
            self.stack.append(x)
            self.min.append(x)
        else:
            self.stack.append(x)
            if x <= self.min[len(self.min)-1]:
                self.min.append(x)

    def pop(self) -> None:
        result = self.stack.pop()
        if result == self.min[len(self.min)-1]:
            self.min.pop()

    def top(self) -> int:
        return self.stack[len(self.stack)-1]

    def getMin(self) -> int:
        return self.min[len(self.min)-1]    


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 0155-min-stack/solution.cpp
class MinStack {
private:
    vector<int> S;
    int min;
public:
    /** initialize your data structure here. */
    MinStack() {
        S = vector<int>();
    }
    
    void push(int x) {
        
        if(S.size() > 0 && x < min)
        {
            min = x;
        }
        else if(S.size() == 0)
        {
            min = x;
        }
        S.push_back(x);

        
    }
    
    void pop() {
        int lastElement = S[S.size() - 1];
        S.pop_back();
        if(S.size() > 0 && lastElement  == min)
        {
            min = S[0];
            for(auto itr = S.begin(); itr != S.end(); itr++)
            {
                if(*itr < min)
                    min = *itr;
            }
        }
        
    }
    
    int top() {
        return S[S.size() - 1];
    }
    
    int getMin() {
        return min;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */
